### PR TITLE
Nav block: remove list markup

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

We're shooting ourselves in the foot by making the navigation block markup overly complex. Using lists in navigation elements is [not even necessary](https://css-tricks.com/navigation-in-lists-to-be-or-not-to-be/). It's not because we've always been doing it that way that it is the correct way. I believe the the navigation links will be announced just fine, and if we need to be more descriptive somewhere, we can always add labels (list count and order comes to mind).

**Rule of thumb: if you have to completely remove list styling, you should probably not be using lists.**

For example, it does make sense to use list if it's a big table of contents of the site, but that's not how this menu will be used I believe.

It's not too late to change since we haven't released the navigation block yet.

While it doesn't really matter if a navigation section uses lists or not, it *does* matter for us because we're making the navigation overly complex by having to manage the lists. Without lists, you can potentially embed *any* block in the navigation block without it needing a list item wrapper.

I just removed the list markup here, but otherwise left the structure intact so I don't need to change CSS. However, with the `ul` element gone, the replacement `div` should probably be removed in a follow-up.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by getdave** [comment]
I understand all your points. From an accessibility viewpoint though are we confident that removing the affordances provided by lists wont provide a poorer UX? For example I believe the total items in lists are announced where as that wouldnt work for more generic markup.

What if the user has a giant navigation? Without that affordance a user of assistive tech might dive into a series of items without any idea of how deep the rabbit hole goes.

Not saying your points arent valid but just checking weve considered all a11y considerations. My knowledge may be out of date but Id rather be safe than sorry.

**Comment by priethor** [comment]
This PR would solve #29036 and open many possibilities. On the other hand, regarding navigation accessibility, as far as I know, it was discussed in [#24364 (comment)](https://github.com/WordPress/gutenberg/issues/24364#issuecomment-670600067) determining links indeed need to be wrapped with a `ul`. It might be worth, though, discussing it again from this perspective.

**Comment by ellatrix** [comment]
> For example I believe the total items in lists are announced where as that wouldnt work for more generic markup.

We should be able to add a label for that.

File: packages/block-library/src/navigation-link/index.php [hunk 1]
```diff
@@ -146,7 +146,7 @@ function render_block_core_navigation_link( $attributes, $content, $block ) {
 			'style' => $style_attribute,
 		)
 	);
-	$html               = '<li ' . $wrapper_attributes . '>' .
+	$html               = '<div ' . $wrapper_attributes . '>' .
```
Comment (gwwar) [inline packages/block-library/src/navigation-link/index.php:149]: Following this line of thinking, I think it should be possible to drop this div wrapper too here in navigation-link and home-link.

This change would need to rework the css for how submenus work, of course.

This is non blocking since changes here work well and now markup validates properly, but I think it'd be worth exploring.

| Example before | Example after |
|-----|-----|
| Screen Shot 2021-05-14 at 10 10 14 AM | Screen Shot 2021-05-14 at 10 24 46 AM |

**Review by gwwar — COMMENTED** [review]
Thanks @ellatrix for making us rethink our existing assumptions! ✨  If I read the threads correctly, this **_assumption_** might sum up to _**we must use a ul for accessible navigation links**_.

On the technical side of things, I do prefer keeping things simple so this approach has a lot of appeal on that front. All types of blocks that we insert into navigation block (including spacer, social links, etc) will now validate as correct html5 markup which is also very good. 

I briefly tested this in VoiceOver and one of the main differences is that UL will list how many links there are. We might be able to work around that by using a good combination of role="navigation" and a good generated `aria-label` to help make up for this.

If anyone has access to the JAWS screen reader I think it'd be good to help spot any other differences here.

The blocking factor for this change I think is whether or not the community here thinks this will work for screen readers. I'd be interested in finding folks who use screen readers daily to provide some feedback, but barring that we might need to bring this up for discussion with a wider group.

So TL;DR, this is very elegant and will 💯 work technically, but we should get feedback on if this is acceptable on the a11y side of things.

File: packages/block-library/src/navigation-link/index.php (hunk 1 above)
Comment (ellatrix) [reply packages/block-library/src/navigation-link/index.php:149]: Definitely, but I didn't want to touch the styling as part of this PR. :)

Comment (gwwar) [reply packages/block-library/src/navigation-link/index.php:149]: Yeah, this could be done as a follow up too if we choose this approach 👍

**Comment by carolinan** [comment]
In the editor, there is still an `UL` left:
https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/edit.js#L176


Disclaimer: I am not a screen reader user, the only time I use a screen reader is for testing. I don't think the usability of the block falls on if a list element is used or not. I found both versions, trunk and this PR, very difficult to use.
With NVDA: 
As with the columns block, the visible focus is lost when I try to move between the menu items.
The visible focus outline stays on the first item, even though the screen reader reads the text link of the second item, and so son.

Where I, as a beginner screen reader user try to move to the next item using tab, I leave the block and end up on the post settings sidebar. Using arrow keys: up and down -nothing happens. Left or right: the text of the current menu item is read out one letter at the time.

Even though all the information below is read out, I am still not able to tell that all menu items have sub menus except for the last item:

With list:
`clickable  Block: Navigation  grouping  list  with 6 items  Block: Custom Link  grouping  Navigation link text  edit  multi line    Pages  out of edit     out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Categories  out of edit     out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Depth  out of edit     out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Advanced  out of edit     out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Menu Description  out of edit  out of grouping  Add block`

With div:
`Block: Navigation  grouping  Block: Custom Link  grouping  Navigation link text  edit  multi line    Pages  out of edit  out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Categories  out of edit  out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Depth  out of edit  out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Advanced  out of edit  out of grouping  Block: Custom Link  grouping  clickable    Navigation link text  edit  multi line    Menu Description  out of edit  out of grouping  Add block
`

**Comment by getdave** [comment]
> clickable Block: Navigation grouping list with 6 items Block:

This is a key example. Imagine if the list had 100 items and it _didn't_ announce the number of items? It would provide a terrible user experience.

We _can_ use labels but 9 times out of 10 the most accessible route is to use plain old HTML. 

I am of course open to the fact that this might be the 1 time HTML doesn't cut the mustard but IMHO we should definitely be bringing this decision up with the a11y teams _before_ we proceed. 

**Update**: I just read [@gwwar's response](https://github.com/WordPress/gutenberg/pull/31827#pullrequestreview-660051962) in which she puts this far more elegantly. I agree with Kerry.

**Comment by draganescu** [comment]
Howdy @ellatrix ! I believe @tellthemachines 's #30430 also tries to free the navigation block from the `ul` container. As far as we grasped from [this conversation](https://wordpress.slack.com/archives/C02RP4X03/p1596813869021900) (Slack), we need though to allow menus to contain lists of links for various compatibility with assitive tech and to enable semantic link lists. It is not clear to me if rooles and labels are good enough.

So, while removing the default UL > LI markup for the [block is good in my opinion](https://github.com/WordPress/gutenberg/issues/23915#issuecomment-659351318), at the same time we need to enable a way to place in the block simple "link list" based navigation.

The problem we're facing is one of UX complexity. We don't know what the initial intent of the user is when a navigation block is used, and how the complexity of the block will evolve as the user adds content. The reason we have a list based markup is only because we did not start with the premise of the navigation as an [open canvas](https://github.com/WordPress/gutenberg/issues/23915#issuecomment-659351318), but more on the pattern of the classic WP menus, which are nested lists of links. For nested lists of links ULs are great. 

Once one starts to imagine the navigation block as a canvas for any kind of menu (including marketing heavy mega menus) the problems of learning curve, initial action etc. immediately pop up. So this is where adding a passthrough block shines as an idea. We make a block like the one in #30430 and the user never knows it's there.

Neither of these solutions (this PR, #30430, a passthrough block, rendering markup on front end only) really solve the problem that by allowing various blocks in menus we're making it a mini editor which is basically a special case of the group block. I think the navigation block is now at a point where the problems it has are not so much about how to implement it but about what do we want it to achieve? I mean [this example](https://github.com/WordPress/gutenberg/pull/30430#issuecomment-829030176) from @jasmussen does it really need to be handled by a single block?

**Comment by ellatrix** [comment]
I'm actually starting to think that lists do fulfil a purpose here. While we could add labels etc., it does nicely group links together as a single entity next to spacers and search blocks, which you probably don't want to mix with links in the menu. For example, it's probably not great if a user could move a search block _between_ links, they should rather only be able to move it before and after the ... _list_.

So with that thought, I'm closing my PR.

**Comment by ellatrix** [comment]
Ok, apparent it _should_ be possible to move non-links into a list of links, such as search or the site title, so in that case I'm again for removing the list markup. I'll add a label for the item count shortly.

**Comment by alexstine** [comment]
Just to confirm, I should be testing the Navigation block after you push the change to add count label?

Thanks.

**Comment by ellatrix** [comment]
@alexstine Yes, I'll add a count label in a bit. I can ping you once I've added it. 🙂

**Comment by joedolson** [comment]
Will there still be a way to define subgroups within the navigation if lists are not used? That's traditionally been done via nested lists, but I'm not clear how this is going to happen without lists. 

Any flow content is permitted within a `<nav>` element; it doesn't have to be restricted to just a list, so it's feasible to do other things. But the average navigation is still probably best represented by a list of links. 

I'm having a little trouble imagining exactly what the end result for this block will be for either the editor or the user, though, so it's hard to comment until after testing. Even then, it sounds like it'll take some serious testing.

**Comment by tellthemachines** [comment]
> Ok, apparent it should be possible to move non-links into a list of links, such as search or the site title

I'm not clear on this point, could you link to sources/discussion? A big part of the a11y concern was isolating links from non-links, so I wouldn't expect it to be possible to move a non-link inside a list of links.

**Comment by jasmussen** [comment]
> I'm not clear on this point, could you link to sources/discussion?

Andrei [summed it up nicely](https://github.com/WordPress/gutenberg/pull/31827#issuecomment-842257962).

Semantic accessible markup is a given. We also need to retain [the current user experience](https://github.com/WordPress/gutenberg/pull/30430#issuecomment-829030176) — additional containers surfaced to the user are not going to be intuitive.

**Comment by jasmussen** [comment]
Should we close this one now the markup change has been addressed elsewhere?

**Comment by ellatrix** [comment]
Sure :)
